File: feature/keytool/src/main/java/net/m3mobile/feature/keytool/internal/KeySettingRequest.kt
package net.m3mobile.feature.keytool.internal

import net.m3mobile.core.device.DeviceModel
import net.m3mobile.feature.keytool.constants.ExtraKey
import net.m3mobile.feature.keytool.constants.KeyToolContract
import net.m3mobile.feature.keytool.constants.RequestAction

internal class KeySettingRequest internal constructor(
    val packageName: String,
    val action: String,
    val ordered: Boolean,
    private val extras: Map<String, Any>,
) {

    fun extra(key: String): Any? = extras[key]

    fun extras(): Map<String, Any> = extras.toMap()

    override fun toString(): String =
        "KeySettingRequest(packageName=$packageName, action=$action, extras=${extras.keys})"
}

internal object KeySettingRequestFactory {

    fun keyFunctionRequest(
        keyName: String,
        function: String,
        wakeUpEnabled: Boolean? = null,
    ): KeySettingRequest {
        require(keyName.isNotBlank()) { "A KeyTool key title is required." }
        require(function.isNotBlank()) { "A KeyTool function title is required." }
        return unifiedRequest(keyName, function, wakeUpEnabled)
    }

    fun scanWakeUpRequest(
        model: DeviceModel,
        keyName: String,
        enabled: Boolean,
    ): KeySettingRequest {
        require(keyName == KeyToolContract.LEFT_SCAN || keyName == KeyToolContract.RIGHT_SCAN) {
            "Unsupported scan key: $keyName"
        }

        return if (model == DeviceModel.SM24) {
            unifiedRequest(keyName, function = null, wakeUpEnabled = enabled)
        } else {
            legacyWakeUpRequest(keyName, enabled)
        }
    }

    private fun unifiedRequest(
        keyName: String,
        function: String?,
        wakeUpEnabled: Boolean?,
    ): KeySettingRequest {
        require(function != null || wakeUpEnabled != null) {
            "At least one KeyTool setting is required."
        }
        val extras = mutableMapOf<String, Any>(ExtraKey.KEY_TITLE to keyName)
        function?.let { extras[ExtraKey.KEY_FUNCTION] = it }
        wakeUpEnabled?.let { extras[ExtraKey.KEY_WAKEUP] = it }

        return KeySettingRequest(
            packageName = KeyToolContract.SL20_PACKAGE,
            action = RequestAction.ACTION_SET_KEY,
            ordered = true,
            extras = extras,
        )
    }

    private fun legacyWakeUpRequest(keyName: String, enabled: Boolean): KeySettingRequest =
        KeySettingRequest(
            packageName = KeyToolContract.LEGACY_WAKE_UP_PACKAGE,
            action = when (keyName) {
                KeyToolContract.LEFT_SCAN -> RequestAction.LEFT_SCAN_WAKEUP
                else -> RequestAction.RIGHT_SCAN_WAKEUP
            },
            ordered = false,
            extras = mapOf(ExtraKey.ENABLE_WAKEUP to enabled),
        )
}

File: dotnet/M3Mobile.M3Sdk.Xamarin/KeyTool/KeySettingRequest.cs
using System;
using System.Collections.Generic;
using M3Sdk.Xamarin.Internal;

namespace M3Sdk.Xamarin.KeyTool
{
    internal sealed class KeySettingRequest
    {
        private readonly IDictionary<string, object> _extras;

        private KeySettingRequest(
            string packageName,
            string action,
            bool ordered,
            IDictionary<string, object> extras)
        {
            PackageName = packageName;
            Action = action;
            Ordered = ordered;
            _extras = new Dictionary<string, object>(extras);
        }

        internal string PackageName { get; private set; }

        internal string Action { get; private set; }

        internal bool Ordered { get; private set; }

        internal IEnumerable<KeyValuePair<string, object>> Extras
        {
            get { return _extras; }
        }

        internal object Extra(string key)
        {
            object value;
            return _extras.TryGetValue(key, out value) ? value : null;
        }

        internal static KeySettingRequest ForKeyFunction(
            string keyName,
            string function,
            bool? wakeUpEnabled = null)
        {
            if (string.IsNullOrWhiteSpace(keyName))
                throw new ArgumentException("A KeyTool key title is required.", nameof(keyName));
            if (string.IsNullOrWhiteSpace(function))
                throw new ArgumentException("A KeyTool function title is required.", nameof(function));

            return Unified(keyName, function, wakeUpEnabled);
        }

        internal static KeySettingRequest ForScanWakeUp(
            DeviceModel model,
            string keyName,
            bool enabled)
        {
            if (keyName != Constants.KeyTool.LeftScan && keyName != Constants.KeyTool.RightScan)
                throw new ArgumentException("Unsupported scan key: " + keyName, nameof(keyName));

            return model == DeviceModel.SM24
                ? Unified(keyName, null, enabled)
                : LegacyWakeUp(keyName, enabled);
        }

        private static KeySettingRequest Unified(
            string keyName,
            string function,
            bool? wakeUpEnabled)
        {
            if (function == null && !wakeUpEnabled.HasValue)
                throw new ArgumentException("At least one KeyTool setting is required.");

            var extras = new Dictionary<string, object>
            {
                { Constants.KeyTool.ExtraKeyTitle, keyName }
            };
            if (function != null)
                extras.Add(Constants.KeyTool.ExtraKeyFunction, function);
            if (wakeUpEnabled.HasValue)
                extras.Add(Constants.KeyTool.ExtraKeyWakeUp, wakeUpEnabled.Value);

            return new KeySettingRequest(
                Constants.KeyTool.Sl20PackageName,
                Constants.KeyTool.SetKey,
                true,
                extras);
        }

        private static KeySettingRequest LegacyWakeUp(string keyName, bool enabled)
        {
            return new KeySettingRequest(
                Constants.KeyTool.WakeUpPackageName,
                keyName == Constants.KeyTool.LeftScan
                    ? Constants.KeyTool.LeftScanWakeUp
                    : Constants.KeyTool.RightScanWakeUp,
                false,
                new Dictionary<string, object>
                {
                    { Constants.KeyTool.ExtraWakeUpEnabled, enabled }
                });
        }

        public override string ToString()
        {
            return "KeySettingRequest(PackageName=" + PackageName +
                   ", Action=" + Action +
                   ", Extras=" + string.Join(", ", _extras.Keys) + ")";
        }
    }
}
